fdbserver/SimpleConfigConsumer.actor.cpp

```diff
@@ -0,0 +1,153 @@
+	Counter successfulChangeRequest;
+	Counter failedChangeRequest;
+	Counter snapshotRequest;
+	Future<Void> logger;
+
+	ACTOR static Future<Void> compactor(SimpleConfigConsumerImpl* self, ConfigBroadcaster* broadcaster) {
+		if (!self->compactionInterval.present()) {
+			wait(Never());
+			return Void();
+		}
+		loop {
+			state Version compactionVersion = self->lastSeenVersion;
+			wait(delayJittered(self->compactionInterval.get()));
+			wait(self->cfi.compact.getReply(ConfigFollowerCompactRequest{ compactionVersion }));
+			++self->compactRequest;
+			broadcaster->compact(compactionVersion);
+		}
+	}
+
+	ACTOR static Future<Void> fetchChanges(SimpleConfigConsumerImpl* self, ConfigBroadcaster* broadcaster) {
+		wait(getSnapshotAndChanges(self, broadcaster));
+		loop {
+			try {
+				ConfigFollowerGetChangesReply reply =
+				    wait(self->cfi.getChanges.getReply(ConfigFollowerGetChangesRequest{ self->lastSeenVersion }));
+				++self->successfulChangeRequest;
+				for (const auto& versionedMutation : reply.changes) {
+					TraceEvent te(SevDebug, "ConsumerFetchedMutation", self->id);
+					te.detail("Version", versionedMutation.version)
+					    .detail("ConfigClass", versionedMutation.mutation.getConfigClass())
+					    .detail("KnobName", versionedMutation.mutation.getKnobName());
+					if (versionedMutation.mutation.isSet()) {
+						te.detail("Op", "Set").detail("KnobValue", versionedMutation.mutation.getValue().toString());
+					} else {
+						te.detail("Op", "Clear");
+					}
+				}
+				if (reply.mostRecentVersion > self->lastSeenVersion) {
+					self->lastSeenVersion = reply.mostRecentVersion;
+					broadcaster->applyChanges(reply.changes, reply.mostRecentVersion, reply.annotations);
+				}
+				wait(delayJittered(self->pollingInterval));
+			} catch (Error& e) {
+				++self->failedChangeRequest;
+				if (e.code() == error_code_version_already_compacted) {
+					wait(getSnapshotAndChanges(self, broadcaster));
+				} else {
+					throw e;
+				}
+			}
+		}
+	}
+
+	ACTOR static Future<Void> getSnapshotAndChanges(SimpleConfigConsumerImpl* self, ConfigBroadcaster* broadcaster) {
+		ConfigFollowerGetSnapshotAndChangesReply reply =
+		    wait(self->cfi.getSnapshotAndChanges.getReply(ConfigFollowerGetSnapshotAndChangesRequest{}));
+		++self->snapshotRequest;
+		TraceEvent(SevDebug, "ConfigConsumerGotSnapshotAndChanges", self->id)
+		    .detail("SnapshotVersion", reply.snapshotVersion)
+		    .detail("SnapshotSize", reply.snapshot.size())
+		    .detail("ChangesVersion", reply.changesVersion)
+		    .detail("ChangesSize", reply.changes.size())
+		    .detail("AnnotationsSize", reply.annotations.size());
+		broadcaster->applySnapshotAndChanges(
+		    std::move(reply.snapshot), reply.snapshotVersion, reply.changes, reply.changesVersion, reply.annotations);
+		self->lastSeenVersion = reply.changesVersion;
+		return Void();
+	}
+
+	static ConfigFollowerInterface getConfigFollowerInterface(ConfigFollowerInterface const& cfi) { return cfi; }
```

Comment:
I guess this is provided as an overloaded function in case `ConfigSource` below is of type `ConfigFollowerInterface`. Seeing this function is a little confusing though 😅

Reply:
Yeah, this is only to avoid having to write two separate `SimpleConfigConsumerImpl` constructors.